File: 015.NN_UIViewAnimationPractice/NN_UIViewAnimationPractice/UIViewAnimationPractice/Views/NNView.m
//
//  NNView.m
//  NN_UIViewAnimationPractice
//
//  Created by 刘朋坤 on 2018/8/11.
//  Copyright © 2018年 刘朋坤. All rights reserved.
//

#import "NNView.h"
#import "NNBottomView.h"

@interface NNView()

/** 底部按钮视图 */
@property (nonatomic, strong) NNBottomView *bottomView;
/** 按钮的数组元素 **/
@property (nonatomic, copy) NSArray *buttonTitleArray;

@end

@implementation NNView

- (instancetype)initWithFrame:(CGRect)frame buttonTitleArray:(NSArray *)buttonTitleArray {
    if (self = [super initWithFrame:frame]) {
        self.backgroundColor = [UIColor cyanColor];
        _buttonTitleArray = buttonTitleArray;
        [self creatViews];
    }
    return self;
}

#pragma mark - 布局区域
- (void)creatViews {
    [self addSubview:self.headerImage];
    [self addSubview:self.bottomView];
}

#pragma mark - 懒加载区域
- (UIImageView *)headerImage {
    if (!_headerImage) {
        _headerImage = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"header"]];
        _headerImage.frame = CGRectMake(100, 100, 200, 200);
        _headerImage.center = self.center;
    }
    return _headerImage;
}

- (NNBottomView *)bottomView {
    if (!_bottomView) {
        _bottomView = [[NNBottomView alloc] initWithFrame:CGRectMake(0, self.frame.size.height - 200, self.frame.size.width, 200) buttonTitleArray:self.buttonTitleArray];
        __weak typeof(self) weakSelf = self;
        _bottomView.actionBlock = ^(id param) {
            if (weakSelf.actionBlock) {
                weakSelf.actionBlock(param);
            }
        };
    }
    return _bottomView;
}


@end

File: 017.NN_UIViewAnimationPractice_swift/NN_UIViewAnimationPractice_swift/UIViewAnimationPractice/Views/NNView.swift
//
//  NNView.swift
//  NN_UIViewAnimationPractice
//
//  Created by 刘朋坤 on 2019/1/5.
//  Copyright © 2019 刘朋坤. All rights reserved.
//

import UIKit

protocol NNViewProtocol: NSObjectProtocol {
    func clickAction(index: NSInteger)
}

class NNView: UIView {
    
    var buttonTitleArray : NSArray = []
    weak var delegate:NNViewProtocol?

    init(frame: CGRect, buttonTitleArray: NSArray) {
        super.init(frame: frame)
        self.buttonTitleArray = buttonTitleArray
        addSubview(headerImage)
        addSubview(bottomView)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // 懒加载区域
    lazy var headerImage : UIImageView = {
        let tempHeaderImage = UIImageView.init(image: UIImage.init(named: "header"))
        tempHeaderImage.frame = CGRect(x: 100, y: 100, width: 200, height: 200)
        tempHeaderImage.center = self.center
        return tempHeaderImage
    }()
    
    lazy var bottomView : NNBottomView = {
        let tempBottomView = NNBottomView.init(frame: CGRect.init(x: 0, y: self.frame.size.height - 200, width: self.frame.size.width, height: 200), buttonTitleArray: buttonTitleArray as NSArray)
        tempBottomView.delegate = self
        return tempBottomView
    }()
    
}

/** 代理区域 - NNBottomViewProtocol */
extension NNView : NNBottomViewProtocol {
    func clickAction(index: NSInteger) {
        if (delegate != nil) {
            delegate?.clickAction(index: index)
        }
    }
}

